src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,368 @@
+
+              print_vector.clear();
+              print_vector.push_back(delta_t);
+              print_vector.push_back(elbo);
+              services::io::write_iteration_csv(*diag_stream_,
+                                                iter_counter, print_vector);
+            }
+
+            if (delta_elbo_ave < tol_rel_obj) {
+              if (print_stream_)
+                *print_stream_ << "   MEAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (delta_elbo_med < tol_rel_obj) {
+              if (print_stream_)
+                *print_stream_ << "   MEDIAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (iter_counter > 100) {
+              if (delta_elbo_med > 0.5 || delta_elbo_ave > 0.5) {
+                if (print_stream_)
+                  *print_stream_ << "   MAY BE DIVERGING... INSPECT ELBO";
+              }
+            }
+
+            if (print_stream_)
+              *print_stream_ << std::endl;
+          }
+
+          // Check for max iterations
+          if (iter_counter == max_iterations) {
+            if (print_stream_)
+              *print_stream_ << "MAX ITERATIONS REACHED" << std::endl;
+            do_more_iterations = false;
+          }
+
+          ++iter_counter;
+        }
+      }
+
+      /**
+       * Runs the algorithm and writes to output.
+       *
+       * @param  tol_rel_obj    relative tolerance parameter for convergence
+       * @param  max_iterations max number of iterations to run algorithm
+       */
+      int run(double tol_rel_obj, int max_iterations) const {
+        if (diag_stream_) {
+          *diag_stream_ << "iter,time_in_seconds,ELBO" << std::endl;
+        }
+
+        // initialize variational approximation
+        Q variational = Q(cont_params_);
+
+        // run inference algorithm
+        robbins_monro_adagrad(variational, tol_rel_obj, max_iterations);
+
+        // get mean of posterior approximation and write on first output line
+        cont_params_ = variational.mean();
+        std::vector<double> cont_vector(cont_params_.size());
+        for (int i = 0; i < cont_params_.size(); ++i)
+          cont_vector.at(i) = cont_params_(i);
+        std::vector<int> disc_vector;
+
+        if (out_stream_) {
+          services::io::write_iteration(*out_stream_, model_, rng_,
+                          0.0, cont_vector, disc_vector, &std::cout);
+        }
+
+        // draw more samples from posterior and write on subsequent lines
+        if (out_stream_) {
+          for (int n = 0; n < n_posterior_samples_; ++n) {
+            cont_params_ = variational.sample(rng_);
+            for (int i = 0; i < cont_params_.size(); ++i)
+              cont_vector.at(i) = cont_params_(i);
+            services::io::write_iteration(*out_stream_, model_, rng_,
+                          0.0, cont_vector, disc_vector, &std::cout);
```

Comment:
should be `print_stream_` instead of `&std::cout`, right? we've removed uses of `std::cout` from code.